sources/mdb_iamproxy.c

```diff
@@ -0,0 +1,254 @@
+			 "failed to connect to %s", exchange_socket.sun_path);
+		authentication_result = MDB_IAMPROXY_CONN_ERROR;
+		goto free_end;
+	}
+
+	/*COMMUNICATE WITH SOCKET*/
+	msg_username = machine_msg_create(0);
+	if (msg_username == NULL) {
+		od_error(&instance->logger, "auth", client, NULL,
+			 "failed to allocate msg_username");
+		authentication_result = MDB_IAMPROXY_CONN_ERROR;
+		goto free_io;
+	}
+	if (machine_msg_write(msg_username, username, strlen(username) + 1) <
+	    0) {
+		od_error(&instance->logger, "auth", client, NULL,
+			 "failed to send username to msg_username");
+		authentication_result = MDB_IAMPROXY_CONN_ERROR;
+		goto free_io;
+	}
+
+	msg_token = machine_msg_create(0);
+	if (msg_token == NULL) {
+		od_error(&instance->logger, "auth", client, NULL,
+			 "failed to allocate msg_token");
+		authentication_result = MDB_IAMPROXY_CONN_ERROR;
+		goto free_io;
+	}
+	if (machine_msg_write(msg_token, token, strlen(token) + 1) < 0) {
+		od_error(&instance->logger, "auth", client, NULL,
+			 "failed to write token to msg_token");
+		authentication_result = MDB_IAMPROXY_CONN_ERROR;
+		goto free_io;
+	}
+
+	correct_sending = mdb_iamproxy_io_write(
+		io, msg_username); // send USERNAME to socket
+	if (correct_sending !=
+	    MDB_IAMPROXY_RES_OK) { // error during sending data to socket
+		od_error(&instance->logger, "auth", client, NULL,
+			 "failed to send username to iam-auth-proxy");
+		authentication_result = correct_sending;
+		goto free_io;
+	}
+	correct_sending =
+		mdb_iamproxy_io_write(io, msg_token); // send TOKEN to socket
+	if (correct_sending !=
+	    MDB_IAMPROXY_RES_OK) { // error during sending data to socket
+		od_error(&instance->logger, "auth", client, NULL,
+			 "failed to send token to iam-auth-proxy");
+		authentication_result = MDB_IAMPROXY_CONN_ERROR;
+		goto free_io;
+	}
+
+	/*COMMUNUCATE WITH SOCKET*/
+	auth_status =
+		mdb_iamproxy_io_read(io); // recieve auth_status from socket
+	if (auth_status == NULL) { // recieving is not completed successfully
+		od_error(&instance->logger, "auth", client, NULL,
+			 "failed to receive auth_status from iam-auth-proxy");
+		authentication_result = MDB_IAMPROXY_CONN_ERROR;
+		goto free_io;
+	}
+
+	auth_status_char = (char *)machine_msg_data(auth_status);
+	if ((unsigned)auth_status_char[0]) {
+		authentication_result = MDB_IAMPROXY_CONN_ACCEPTED;
+	} else {
+		authentication_result = MDB_IAMPROXY_CONN_DENIED;
+	}
+
+	external_user =
+		mdb_iamproxy_io_read(io); // recieve subject_id from socket
+	if (external_user == NULL) {
+		od_error(&instance->logger, "auth", client, NULL,
+			 "failed to receive external_user from iam-auth-proxy");
+		authentication_result = MDB_IAMPROXY_CONN_ERROR;
+		goto free_auth_status;
+	}
+
+	od_log(&instance->logger, "auth", client, NULL,
+	       "user '%s.%s' was authenticated with subject_id: %s",
+	       client->startup.database.value, client->startup.user.value,
+	       (char *)machine_msg_data(external_user));
```

Comment:
I guess we need to store `external_user` in `client` somehow and use it for structured logging